### ISSUE TITLE
ClassDeclaration - Add "Extract Interface"

### ISSUE THREAD

**Issue by dsherret** [dsherret/ts-simple-ast#469]
It would be useful to be able to extract an interface structure from a class declaration.

Probably something like this:

```ts
extractInterface(name?: string): InterfaceStructure;
```

**Comment by cancerberoSgx** [comment]
nice! go structures! :) I think there's no problem with having unused imports since if users are trying to build a sourcefile for interfaces they can always execute organizeImports. I was thinking about suggesting to add "exclude" or "include" member names since for big classes users will probably want to split it into several implemented interfaces but then I thought, KISS, I can filter the resulting structure members. I know supporting structures for such a language (that release new features often) could be a hard work and it will be always incomplete, but I think it's a killer feature, thanks!

**Comment by cancerberoSgx** [comment]
I agree with that structures shouldn't be polluted with dependencies. But I can see where you want to go with structures as refactoring tools . IMO structures shouldn't know anything about type binding or execution context. Maybe we need to introduce another concept for representing dependencies - BTW I'm just working on a related problem - I want to implement a "move declaration to another file" refactor - I'm going slowly but I think I'm making progress. 

The main algorithm:
https://github.com/cancerberoSgx/typescript-plugins-of-mine/blob/master/typescript-plugin-move-declaration/src/moveNode.ts

The implementation: 
https://github.com/cancerberoSgx/typescript-plugins-of-mine/blob/master/typescript-plugin-move-declaration/src/moveNodeUtil.ts

Specs: 
https://github.com/cancerberoSgx/typescript-plugins-of-mine/tree/master/typescript-plugin-move-declaration/spec

Mainly I'm using structures to move the nodes, and for dependencies - that's the difficult part, i'm using node.findReferences(),  sourceFile.getRelativePathAsModuleSpecifierTo for creating new import declarations. In the Util.ts file there's a small findReferencesDeclaredOutside() function that I use to create import declarations of those nodes that are not inside the node to move but not imported - and those I need to make exportble and import then in the target file. So far so good, thanks to ts-simple-ast, for small projects it's working almost fine. I tried to use it in the ts-simple-ast source code project and there it won't behave too well, but I think I'm progressing. 

BTW2 I implemented an extract interface plugin that more or less worked OK, also based on tsa, but now that you implement yours probably will work better - so probably I will change my implementation to use your API. 

I had implemented several of this kind of refactors - I can share with you some ideas - research - but thinking on the developer using a keyboard not thinking on an API like you are doing. It's interesting. 

Thanks, keep it up

**Comment by dsherret** [comment]
@cancerberoSgx yeah, I agree with what you said.

Cool! That's an interesting idea. For moving a declaration to another file that might get complicated... for example, what should happen in the following scenario?

```ts
interface Test {
    prop: string;
}
interface Test {
    name: string;
}

export { Test };
```

Or in this scenario?

```ts
const someVar = 5; // should this be moved too?

export function test() {
    return someVar;
}
```

Anyway, something to think about.

**Comment by cancerberoSgx** [comment]
In the second scenario what I'm diong is only moving test and making someVar exportable. The first one is a little bit tricky - what I'm doing is not doing anything is there are compilation errors. This is another complicated scenario: moving a declaration that is default export to a file that already has a default export. In that case I'm just aborting.  But yes, you are right, is not a simple problem - still working on it
